File: c/algorithms/counting.c
#include "counting.h"
#include <stdlib.h>

/*
Counting Sort (stable)
Steps:
1) Handle trivial input (length < 2).
2) Find min and max to compute the range.
3) Allocate count array of size R = max - min + 1, initialized to zero.
4) Count each element's frequency (offset by min so negatives work).
5) Prefix sums over count array → cumulative counts (final positions).
6) Allocate output array and fill it stably (scan input from right to left).
7) Copy output back into arr and free temp arrays.
*/
void counting_sort(int* arr, size_t length) {
    if (length < 2) return;

    int min = arr[0], max = arr[0];
    for (size_t i = 1; i < length; i++) {
        if (arr[i] < min) min = arr[i];
        if (arr[i] > max) max = arr[i];
    }

    size_t range = (size_t)(max - min + 1);
    int* count = (int*)calloc(range, sizeof(int));
    if (!count) return;

    for (size_t i = 0; i < length; i++) {
        count[arr[i] - min]++;
    }

    for (size_t i = 1; i < range; i++) {
        count[i] += count[i - 1];
    }

    int* output = (int*)malloc(length * sizeof(int));
    if (!output) { free(count); return; }

    for (size_t i = length; i-- > 0;) {
        int key = arr[i] - min;
        count[key]--;
        output[count[key]] = arr[i];
    }

    for (size_t i = 0; i < length; i++) {
        arr[i] = output[i];
    }

    free(count);
    free(output);
}

File: python/algorithms/counting.py
def counting_sort(arr):
    """
    Stable counting sort for integers (supports negatives).

    Steps:
      1) Handle trivial input (0/1 item).
      2) Find global min/max to compute the value RANGE.
      3) Create an empty "count" array of size R = max - min + 1 (all zeros).
      4) Tally occurrences: count[x - min]++ for each x in arr.
      5) Prefix sums over count -> positions: count[i] accumulates elements <= i.
      6) Build output stably by scanning arr from RIGHT to LEFT.
      7) Copy output back into arr (in-place for callers).
    """
    n = len(arr)
    if n < 2:
        return

    mn = min(arr)
    mx = max(arr)
    R = mx - mn + 1

    # (3) index array initialized to zero
    count = [0] * R

    # (4) frequency tally (offset by mn so negatives work)
    for x in arr:
        count[x - mn] += 1

    # (5) prefix sums -> stable placement indices
    for i in range(1, R):
        count[i] += count[i - 1]

    # (6) stable fill of output by going backwards
    out = [0] * n
    for i in range(n - 1, -1, -1):
        k = arr[i] - mn
        count[k] -= 1                 # last free slot for this key
        out[count[k]] = arr[i]        # place arr[i]

    # (7) write back into the caller's list
    arr[:] = out
